src/protocol/sort/reshare.rs

```diff
@@ -0,0 +1,138 @@
+    field::Field,
+    helpers::{
+        mesh::{Gateway, Mesh},
+        prss::SpaceIndex,
+        Direction, Identity,
+    },
+    protocol::{context::ProtocolContext, RecordId, Step},
+    secret_sharing::Replicated,
+};
+use embed_doc_image::embed_doc_image;
+use serde::{Deserialize, Serialize};
+
+/// A message sent by each helper when they've reshared their own shares
+#[derive(Debug, Serialize, Deserialize, Default)]
+pub struct PartValue<F> {
+    part: F,
+}
+/// Reshare(i, [x])
+// This implements reshare algorithm of "Efficient Secure Three-Party Sorting Protocol with an Honest Majority" at communication cost of 2R.
+// Input: Pi-1 and Pi+1 knows their secret shares
+// Output: At the end of the protocol, [all 3 helpers receive renewed shares of x with randomness added
+#[derive(Debug)]
+pub struct Reshare<F> {
+    input: Replicated<F>,
+}
+
+impl<F: Field> Reshare<F> {
+    #[allow(dead_code)]
+    pub fn new(input: Replicated<F>) -> Self {
+        Self { input }
+    }
+
+    /// Steps
+    #[embed_doc_image("reshare", "images/sort/reshare.png")]
+    /// 1. While calculating for helper, we call PRSS to get randoms for helper who needs to receive reshares (say `rand_left`, `rand_right`)
+    ///    `to_helper.left` knows `rand_left` (named r1) and `to_helper.right` knows `rand_right` (named r0)
+    /// 2. `to_helper.left` calculates part1 = (input.0 + input.1) - r1
+    ///    `to_helper.right` calculates part2 = (input.0 - r0)
+    /// 3. `to_helper.left` and `to_helper.right` exchange their calculated shares
+    /// 4. Everyone sets their shares  
+    ///    `to_helper.left`  = (part1 + part2, `rand_left`)  = (part1 + part2, r1)
+    ///    `to_helper`       = (`rand_left`, `rand_right`)     = (r0, r1)
+    ///    `to_helper.right` = (`rand_right`, part1 + part2) = (r0, part1 + part2)
+    #[allow(dead_code)]
+    pub async fn execute<M: Mesh, G: Gateway<M, S>, S: Step + SpaceIndex>(
+        self,
+        ctx: &ProtocolContext<'_, G, S>,
+        record_id: RecordId,
+        step: S,
+        to_helper: Identity,
+    ) -> Result<Replicated<F>, BoxError> {
+        let mut channel = ctx.gateway.get_channel(step);
+        let prss = &ctx.participant[step];
+        let (r0, r1) = prss.generate_fields(record_id.into());
+
+        let inputs = self.input.as_tuple();
+        // `to_helper.left` calculates part1 = (input.0 + input.1) - r1 and sends part1 to `to_helper.right`
+        if channel.identity() == to_helper.peer(Direction::Left) {
+            let part1 = inputs.0 + inputs.1 - r1;
+            channel
+                .send(
+                    to_helper.peer(Direction::Right),
+                    record_id,
+                    PartValue { part: part1 },
+                )
+                .await?;
+
+            // Sleep until `to_helper.left` sends us their part2 value
+            let PartValue { part: part2 } = channel
+                .receive(channel.identity().peer(Direction::Left), record_id)
```

Comment:
Wait a minute... don't you mean RIGHT? In this fork of the code, we *are* the "left" and we are now waiting to get part2 from the *right*, correct?

Reply:
Nope, the protocol does not communicate at all with the to_helper node. eg if we are computing for H2, H1 sends its parts to H3 and H3 sends parts to H1! H2 just assigns the prss random numbers as its' shares.